datumaro/components/project.py

```diff
@@ -2,486 +2,2389 @@
+
+    def get_subset(self, name):
+        return self._dataset.get_subset(name)
+
+    def categories(self):
+        return self._dataset.categories()
+
+    def get(self, id, subset=None):
+        return self._dataset.get(id, subset)
+
+
+class IgnoreMode(Enum):
+    rewrite = auto()
+    append = auto()
+    remove = auto()
+
+def _update_ignore_file(paths: Union[str, List[str]], repo_root: str,
+        filepath: str, mode: Optional[IgnoreMode] = None):
+    def _make_ignored_path(path):
+        path = osp.join(repo_root, osp.normpath(path))
+        assert path.startswith(repo_root), path
+        return osp.relpath(path, repo_root)
+
+    header = '# The file is autogenerated by Datumaro'
+
+    mode = parse_str_enum_value(mode, IgnoreMode, IgnoreMode.append)
+
+    if isinstance(paths, str):
+        paths = [paths]
+    paths = {osp.join(repo_root, osp.normpath(p)): _make_ignored_path(p)
+        for p in paths}
+
+    openmode = 'r+'
+    if not osp.isfile(filepath):
+        openmode = 'w+' # r+ cannot create, w+ truncates
+    with open(filepath, openmode) as f:
+        lines = []
+        if mode in {IgnoreMode.append, IgnoreMode.remove}:
+            for line in f:
+                lines.append(line.strip())
+            f.seek(0)
+
+        new_lines = []
+        for line in lines:
+            if not line or line.startswith('#'):
+                new_lines.append(line)
+                continue
+
+            line_path = osp.join(repo_root,
+                osp.normpath(line.split('#', maxsplit=1)[0]).lstrip('/'))
+
+            if mode == IgnoreMode.append:
+                if line_path in paths:
+                    paths.pop(line_path)
+                new_lines.append(line)
+            elif mode == IgnoreMode.remove:
+                if line_path in paths:
+                    paths.pop(line_path)
+                else:
+                    new_lines.append(line)
+
+        if mode in { IgnoreMode.rewrite, IgnoreMode.append }:
+            new_lines.extend(paths.values())
+
+        if not new_lines or new_lines[0] != header:
+            f.write(header + '\n')
+        f.writelines('\n'.join(new_lines))
+        f.truncate()
+
+class CrudProxy:
+    @property
+    def _data(self):
+        raise NotImplementedError()
+
+    def __len__(self):
+        return len(self._data)
+
+    def __getitem__(self, name):
+        return self._data[name]
+
+    def get(self, name, default=None):
+        return self._data.get(name, default)
+
+    def __iter__(self):
+        return iter(self._data.keys())
+
+    def items(self):
+        return iter(self._data.items())
+
+    def __contains__(self, name):
+        return name in self._data
+
+class _DataSourceBase(CrudProxy):
+    def __init__(self, project, config_field):
         self._project = project
-        self._env = project.env
-        config = self.config
-        env = self.env
+        self._field = config_field
+
+    @CrudProxy._data.getter
```

Comment:
This is a bit confusing; doesn't it essentially redefine the `_data` property _in_ the `CrudProxy` class? I would've expected to just see `@property` here.

Reply:
It's just how it works https://stackoverflow.com/a/7020531. I don't like it either.

Reply:
That question is about overriding a specific component of a property (e.g. just the getter). It's not really applicable here, since `_data` is a readonly property.

Reply:
Ok, changed.

Reply:
There's another instance of this in `ProjectBuildTargets`.

Reply:
Fixed